compaction/compaction_manager.cc

```diff
@@ -1267,19 +1271,23 @@ class regular_compaction_task_executor : public compaction_task_executor, public
 
 }
 
-void compaction_manager::submit(table_state& t) {
+void compaction_manager::submit(table_state& t, std::optional<std::vector<sstables::shared_sstable>> candidates_opt) {
     if (t.is_auto_compaction_disabled_by_user()) {
         return;
     }
 
+    if (candidates_opt && candidates_opt->empty()) {
+        return;
+    }
+
```

Comment:
this is candidates for the compaction strategy.

but who evaluates the need for a task is the compaction strategy.

could happen that a strategy is satisfied but candidates_opt will not be empty, and we'd fall into the same problem.

Additionally, we went a long way to allow compaction strategy to receive either run or raw sstables, so sstables are today only fed into strategy by control. That was to avoid expensive cpu cycles to convert sstables into runs, in ICS.

What I believe is the proper way to tackle this problem is to decouple compaction_manager task from task_manager task (for regular compaction). We cannot avoid creating a compaction_manager task (as we have to wait for an async task on stop()), but we can avoid creating a task_manager one. Every time compaction_manager task decides to run a regular compaction, then we create a task_manager task for it.

Reply:
Why can't we avoid creating a compaction_manager task?
Say the compaction strategy would expose a heuristic function that can tell if definitely has nothing to do synchronously, with no allocations, couldn't we use it in `submit` to avoid the expensive path if it's not required?

Reply:
By the way, @raphaelsc, I see that `strategy_control::candidates_as_runs` is currently used only by tests.

Reply:
I'll drop the last patch from this series so we can expedite the other patches.
So we can continue the discussion over a separate PR.

Reply:
> By the way, @raphaelsc, I see that `strategy_control::candidates_as_runs` is currently used only by tests.

I dropped the ball on the PR that wires it into ICS :-( It's a important patch, I'll revive it today.

Reply:
I think you're right @bhalevy. We can extract all the code for evaluation, including check if we can run the compaction (based on the weight of its descriptor), the registration of compacting, and forward all of that to regular_compaction_task.

Then regular_compaction_task, can use a do while {}. do is guaranteed to execute a task, whereas only next iterations will reevaluate again.

Reply:
@bhalevy one last item that was in my mind, we probably still need to run the evaluation code under compaction sched group, to not interfere with foreground activity. but we can easily do it with a background fiber, e.g.:

`submit() { with_gate(...) { switch_to(compaction_sched); auto ret = run_reevaluation_code(); if (ret) { create_task(ret); } } }`

Reply:
> @bhalevy one last item that was in my mind, we probably still need to run the evaluation code under compaction sched group, to not interfere with foreground activity. but we can easily do it with a background fiber, e.g.:
> 
> `submit() { with_gate(...) { switch_to(compaction_sched); auto ret = run_reevaluation_code(); if (ret) { create_task(ret); } } }`

I'm not sure exactly where this would fit.

Reply:
something like:

```
void compaction_manager::submit(table_state& t) {
    ...

    (void) with_gate(...) { 
        switch_to(compaction_sched);
        
        optional<struct whatever> ret = reevaluate_regular_compaction_for(t);
        
        if (ret) { 
            co_await perform_compaction<regular_compaction_task_executor>(..., std::move(ret));
        }
     }
}
```

Reply:
ok, that makes more sense now.

I just submitted v2, without the patch changing submit.
Let's work on that as a follow up.